File: repositories/productsRepo.js
let products = [];
let nextId = 1;

const add = (data) => {
  const p = {
    id: nextId++,
    name: data.name,
    price: data.price,
    stock: data.stock
  };
  products.push(p);
  return p;
};

const getAll = () => products.map(p => ({ id: p.id, name: p.name, price: p.price, stock: p.stock }));

const getById = (id) => products.find(p => p.id === id) || null;

const update = (id, data) => {
  const p = products.find(x => x.id === id);
  if (!p) return null;
  if (data.name !== undefined) p.name = data.name;
  if (data.price !== undefined) p.price = data.price;
  if (data.stock !== undefined) p.stock = data.stock;
  return p;
};

const decreaseStock = (id, qty) => {
  const p = products.find(x => x.id === id);
  if (!p) return false;
  p.stock -= qty;
  return true;
};

module.exports = { add, getAll, getById, update, decreaseStock };

File: products-service/app/repositories/products_repo.py
class ProductsRepo:
    def __init__(self):
        self.products = []
        self.next_id = 1

    def add(self, data):
        if data.get("price") is None or data.get("stock") is None:
            raise ValueError("price and stock are required")
        product = {
            "id": self.next_id,
            "name": data.get("name"),
            "price": data.get("price"),
            "stock": data.get("stock")
        }
        self.products.append(product)
        self.next_id += 1
        return product

    def get_all(self):
        return self.products

    def get_by_id(self, id):
        return next((p for p in self.products if p["id"] == id), None)

    def update(self, id, data):
        p = self.get_by_id(id)
        if not p:
            return None
        if "name" in data: p["name"] = data["name"]
        if "price" in data: p["price"] = data["price"]
        if "stock" in data: p["stock"] = data["stock"]
        return p

    def decrease_stock(self, id, quantity):
        p = self.get_by_id(id)
        if not p:
            return None
        if p["stock"] < quantity:
            return None
        p["stock"] -= quantity
        return p
